File: resources/loaders/Sprite Kit/WorldLoader.m
//
//  b2dWorldLoader.m
//
//  Created by Amit Kumar Mehar on 19/07/15.
//  Copyright (c) 2015 Amit Kumar Mehar. All rights reserved.
//

#import <Foundation/Foundation.h>

#import "WorldLoader.h"

@implementation WorldLoader

@synthesize offsetX, offsetY;
@synthesize loadedBodies, loadedJoints;

+(id) init{
    
    return self;
}

-(void) reset{
    // release arrays
    if ([loadedBodies count] > 0){
        [loadedBodies removeAllObjects];
        loadedBodies = nil;
    }
    if ([loadedJoints count] > 0){
        [loadedJoints removeAllObjects];
        loadedJoints = nil;
    }
    
    // reinitialize array
    loadedBodies = [[NSMutableArray alloc] init];
    loadedJoints = [[NSMutableArray alloc] init];
}

-(void) loadJsonSceneFromFile:(NSString *)file : (SKScene*)scene{
    [self reset];
    
    // load the json file
    NSString *filePath = [[NSBundle mainBundle] pathForResource:file ofType:@"json"];
    NSString *data = [[NSString alloc] initWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:NULL];
    
    // parse the loaded file to json object
    NSDictionary *jsonScene = [data objectFromJSONString];
    
    // load physics bodies
    NSArray *jsonBodies = [jsonScene objectForKey:@"bodies"];
    for (int i = 0, numberOfBodies = [jsonBodies count]; i < numberOfBodies; i++){
        // load body's properties
        NSDictionary *jsonBody = [jsonBodies objectAtIndex:i];
        NSArray *position = [jsonBody objectForKey:@"position"];
        float rotation = -[[jsonBody objectForKey:@"rotation"] floatValue] * M_PI / 180.0,
            linearDamping = [[jsonBody objectForKey:@"linearDamping"] floatValue],
            angularDamping = [[jsonBody objectForKey:@"angularDamping"] floatValue];
        bool isFixedRotation = [[jsonBody objectForKey:@"isFixedRotation"] boolValue];
        int bodyType = [[jsonBody objectForKey:@"type"] integerValue];
        NSString *userData = [jsonBody objectForKey:@"userData"];
        
        // create node for body
        SKNode *body = [[SKNode alloc] init];
        [body setName: userData];
        [body setPosition:CGPointMake(offsetX + [[position objectAtIndex:0] floatValue], offsetY - [[position objectAtIndex:1] floatValue])];
        [body setZRotation:rotation];
        [scene addChild:body];
        
        // load fixtures / shapes for the body
        NSArray *jsonFixtures = [jsonBody objectForKey:@"fixtures"];
        // add fixtures to the body (also initializes physics body for current body node)
        [self loadJsonFixtures:jsonFixtures :body];
        
        if (bodyType == 0){
            [body.physicsBody setDynamic:false];
        }
        // kinematic bodies are not supported in sprite kit
        else if (bodyType == 2){
            [body.physicsBody setDynamic:true];
        }
        [body.physicsBody setLinearDamping:linearDamping];
        [body.physicsBody setAngularDamping:angularDamping];
        [body.physicsBody setAllowsRotation:!isFixedRotation];
        
        [loadedBodies addObject:body.physicsBody];
    }
    
    // load physics joints
    NSArray *jsonJoints = [jsonScene objectForKey:@"joints"];
    for (int i = 0, numberOfJoints = [jsonJoints count]; i < numberOfJoints; i++){
        NSDictionary *jsonJoint = [jsonJoints objectAtIndex:i];
        
        int jointType = [[jsonJoint objectForKey:@"jointType"] integerValue];
        
        if (jointType == JOINT_DISTANCE){
            int indexA = [[jsonJoint objectForKey:@"bodyA"] integerValue];
            int indexB = [[jsonJoint objectForKey:@"bodyB"] integerValue];
            SKPhysicsBody *bodyA = [loadedBodies objectAtIndex:indexA];
            SKPhysicsBody *bodyB = [loadedBodies objectAtIndex:indexB];
            CGPoint anchorA = CGPointMake([[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:1] floatValue]);
            anchorA.x += bodyA.node.position.x;
            anchorA.y += bodyA.node.position.y;
            CGPoint anchorB = CGPointMake([[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:1] floatValue]);
            anchorB.x += bodyB.node.position.x;
            anchorB.y += bodyB.node.position.y;
            float dampingRatio = [[jsonJoint objectForKey:@"dampingRatio"] floatValue];
            float frequency = [[jsonJoint objectForKey:@"frequencyHZ"] floatValue];
            
            SKPhysicsJointSpring *joint = [SKPhysicsJointSpring jointWithBodyA:bodyA bodyB:bodyB anchorA:anchorA anchorB:anchorB];
            [joint setDamping:dampingRatio];
            [joint setFrequency:frequency];
            [scene.physicsWorld addJoint:joint];
            
            [loadedJoints addObject:joint];
        }
        else if (jointType == JOINT_WELD){
            int indexA = [[jsonJoint objectForKey:@"bodyA"] integerValue];
            int indexB = [[jsonJoint objectForKey:@"bodyB"] integerValue];
            SKPhysicsBody *bodyA = [loadedBodies objectAtIndex:indexA];
            SKPhysicsBody *bodyB = [loadedBodies objectAtIndex:indexB];
            CGPoint anchorA = CGPointMake([[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:1] floatValue]);
            anchorA.x += bodyA.node.position.x;
            anchorA.y += bodyA.node.position.y;
            CGPoint anchorB = CGPointMake([[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:1] floatValue]);
            anchorB.x += bodyB.node.position.x;
            anchorB.y += bodyB.node.position.y;
            float referenceAngle = -[[jsonJoint objectForKey:@"referenceAngle"] floatValue] * M_PI / 180.0;
            
            SKPhysicsJointFixed *joint = [SKPhysicsJointFixed jointWithBodyA:bodyA bodyB:bodyB anchor:anchorB];
            [scene.physicsWorld addJoint:joint];
            
            [loadedJoints addObject:joint];
        }
        else if (jointType == JOINT_REVOLUTE){
            int indexA = [[jsonJoint objectForKey:@"bodyA"] integerValue];
            int indexB = [[jsonJoint objectForKey:@"bodyB"] integerValue];
            SKPhysicsBody *bodyA = [loadedBodies objectAtIndex:indexA];
            SKPhysicsBody *bodyB = [loadedBodies objectAtIndex:indexB];
            CGPoint anchorA = CGPointMake([[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:1] floatValue]);
            anchorA.x += bodyA.node.position.x;
            anchorA.y += bodyA.node.position.y;
            CGPoint anchorB = CGPointMake([[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:1] floatValue]);
            anchorB.x += bodyB.node.position.x;
            anchorB.y += bodyB.node.position.y;
            bool enableLimit = [[jsonJoint objectForKey:@"enableLimit"] boolValue];
            float upperAngle = -[[jsonJoint objectForKey:@"lowerAngle"] floatValue] * M_PI / 180.0;
            float lowerAngle = -[[jsonJoint objectForKey:@"upperAngle"] floatValue] * M_PI / 180.0;
            float referenceAngle = -[[jsonJoint objectForKey:@"referenceAngle"] floatValue] * M_PI / 180.0;
            bool enableMotor = [[jsonJoint objectForKey:@"enableMotor"] boolValue];
            float motorSpeed = -[[jsonJoint objectForKey:@"motorSpeed"] floatValue] * M_PI / 180;
            float maxMotorTorque = [[jsonJoint objectForKey:@"maxMotorTorque"] floatValue];
            
            SKPhysicsJointPin *joint = [SKPhysicsJointPin jointWithBodyA:bodyA bodyB:bodyB anchor:anchorB];
            [joint setShouldEnableLimits:enableLimit];
            [joint setUpperAngleLimit:upperAngle];
            [joint setLowerAngleLimit:lowerAngle];
            [joint setRotationSpeed:motorSpeed];
            [scene.physicsWorld addJoint:joint];
            
            [loadedJoints addObject:joint];
        }
        else if (jointType == JOINT_ROPE){
            int indexA = [[jsonJoint objectForKey:@"bodyA"] integerValue];
            int indexB = [[jsonJoint objectForKey:@"bodyB"] integerValue];
            SKPhysicsBody *bodyA = [loadedBodies objectAtIndex:indexA];
            SKPhysicsBody *bodyB = [loadedBodies objectAtIndex:indexB];
            CGPoint anchorA = CGPointMake([[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorA"] objectAtIndex:1] floatValue]);
            anchorA.x += bodyA.node.position.x;
            anchorA.y += bodyA.node.position.y;
            CGPoint anchorB = CGPointMake([[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:0] floatValue], -[[[jsonJoint objectForKey:@"localAnchorB"] objectAtIndex:1] floatValue]);
            anchorB.x += bodyB.node.position.x;
            anchorB.y += bodyB.node.position.y;
            float maxLength = [[jsonJoint objectForKey:@"maxLength"] floatValue];
            
            SKPhysicsJointLimit *joint = [SKPhysicsJointLimit jointWithBodyA:bodyA bodyB:bodyB anchorA:anchorA anchorB:anchorB];
            [joint setMaxLength:maxLength];
            [scene.physicsWorld addJoint:joint];
            
            [loadedJoints addObject:joint];
        }
    }
}

-(void) loadJsonFixtures : (NSArray*)jsonFixtures : (SKNode*) parent{
    NSMutableArray *bodies = [[NSMutableArray alloc] init];
    
    for (int i = 0, numberOfFixtures = [jsonFixtures count]; i < numberOfFixtures; i++){
        
        // load fixture data
        NSDictionary* jsonFixture = [jsonFixtures objectAtIndex:i];
        float density = [[jsonFixture objectForKey:@"density"] floatValue],
            friction = [[jsonFixture objectForKey:@"friction"] floatValue],
            restitution = [[jsonFixture objectForKey:@"restitution"] floatValue];
        int maskBits = [[jsonFixture objectForKey:@"maskBits"] integerValue],
            categoryBits = [[jsonFixture objectForKey:@"categorykBits"] integerValue],
            groupIndex = [[jsonFixture objectForKey:@"groupIndex"] integerValue];
        
        // load shapes and create physics body
        NSArray *jsonShapes = [jsonFixture objectForKey:@"shapes"];
        for (int j = 0, numberOfShapes = [jsonShapes count]; j < numberOfShapes; j++){
            NSDictionary *jsonShape = [jsonShapes objectAtIndex:j];
            
            NSArray *position = [jsonShape objectForKey:@"position"];
            CGPoint pos = CGPointMake([[position objectAtIndex:0] floatValue], -[[position objectAtIndex:1] floatValue]);
            
            SKPhysicsBody *body;
            
            int shapeType = [[jsonShape objectForKey:@"type"] integerValue];
            if (shapeType == SHAPE_BOX){
                float width = [[jsonShape objectForKey:@"width"] floatValue];
                float height = [[jsonShape objectForKey:@"height"] floatValue];
                body = [SKPhysicsBody bodyWithRectangleOfSize:CGSizeMake(width, height) center: pos];
            }
            else if (shapeType == SHAPE_CIRCLE){
                float radius = [[jsonShape objectForKey:@"radius"] floatValue];
                body = [SKPhysicsBody bodyWithCircleOfRadius:radius * 2 center:pos];
            }
            else if (shapeType == SHAPE_POLYGON){
                NSArray *jsonVerts = [jsonShape objectForKey:@"vertices"];
                
                CGMutablePathRef path = CGPathCreateMutable();
                for (int k = 0, numberOfVertices = [jsonVerts count]; k < numberOfVertices; k++){
                    float x = pos.x + [[[jsonVerts objectAtIndex:k] objectAtIndex:0] floatValue];
                    float y = pos.y - [[[jsonVerts objectAtIndex:k] objectAtIndex:1] floatValue];
                    
                    if (k == 0){
                        CGPathMoveToPoint(path, NULL, x, y);
                    }
                    else {
                        CGPathAddLineToPoint(path, NULL, x, y);
                    }
                    
                }
                
                body = [SKPhysicsBody bodyWithPolygonFromPath:path];
            }
            else if (shapeType == SHAPE_CHAIN){
                NSArray *jsonVerts = [jsonShape objectForKey:@"vertices"];
                
                CGMutablePathRef path = CGPathCreateMutable();
                for (int k = 0, numberOfVertices = [jsonVerts count]; k < numberOfVertices; k++){
                    float x = pos.x + [[[jsonVerts objectAtIndex:k] objectAtIndex:0] floatValue];
                    float y = pos.y - [[[jsonVerts objectAtIndex:k] objectAtIndex:1] floatValue];
                    
                    if (k == 0){
                        CGPathMoveToPoint(path, NULL, x, y);
                    }
                    else {
                        CGPathAddLineToPoint(path, NULL, x, y);
                    }
                    
                }
                
                body = [SKPhysicsBody bodyWithEdgeChainFromPath:path];
            }
            
            [body setDensity:density];
            [body setFriction:friction];
            [body setRestitution:restitution];
            
//            [body setCategoryBitMask:categoryBits];
//            [body setCollisionBitMask:maskBits];
//            [body setContactTestBitMask:groupIndex];
            
            [bodies addObject:body];
            
        }
        
    }
    
    parent.physicsBody = [SKPhysicsBody bodyWithBodies:bodies];
}

@end

File: resources/loaders/Unity3D/WorldLoader.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;	// for List<>
using SimpleJSON;					// for parsing json file

public class WorldLoader : MonoBehaviour{

	/** meter to pixel ratio */
	public float RATIO = 30;

	/** scene to load, file must be present in resources folder **/
	public Object scene;

	private enum ShapeTypes {
		SHAPE_BOX,			// BoxCollider2D
		SHAPE_CIRCLE,		// CircleCollider2D
		SHAPE_POLYGON,		// PolygonCollider2D
		SHAPE_CHAIN			// EdgeCollider2D
	};
	private enum JointTypes {
		JOINT_DISTANCE,		// created as DistanceJoint2D
		JOINT_WELD,			// not supported
		JOINT_REVOLUTE,		// created as HingeJoint2D
		JOINT_WHEEL,		// created as WheelJoint2D
		JOINT_PULLEY,		// not supported
		JOINT_GEAR, 		// not supported
		JOINT_PRISMATIC,	// created as SliderJoint2D 
		JOINT_ROPE			// created as DistanceJoint2D
	};

	// list of loaded gameobjects
	[SerializeField]	// so that entering play mode does not screw up "delete scene" feature 
	[HideInInspector]	// so that user does not screw with loaded objects
	private List<GameObject> loadedObjects;

	// list of loaded joints
	[SerializeField] 
	[HideInInspector]
	private List<Joint2D> loadedJoints;

	// resets loader to load new scene
	public void reset(){
		if (loadedObjects != null){
			loadedObjects.Clear();
			loadedObjects = null;
		}
		loadedObjects = new List<GameObject>();

		if (loadedJoints != null){
			loadedJoints.Clear();
			loadedJoints = null;
		}
		loadedJoints = new List<Joint2D>();
	}

	public List<GameObject> getLoadedObjects(){
		return loadedObjects;
	}

	public List<Joint2D> getLoadedJoints(){
		return loadedJoints;
	}

	public void loadJsonScene(){
		reset();
	
		// load file
		TextAsset file = (TextAsset) Resources.Load(scene.name);
		// parse json file to scene
		JSONNode jsonScene = JSON.Parse(file.text);

		// load bodies
		JSONNode jsonBodies = jsonScene["bodies"];
		loadJsonBodies(jsonBodies);

		// load joints
		JSONNode jsonJoints = jsonScene["joints"];
		loadJsonJoints(jsonJoints);
	}

	void loadJsonJoints(JSONNode jsonJoints){
		int JointCount = 0;

		for (int i = 0, numberOfJoints = jsonJoints.Count; i < numberOfJoints; i++){
			JSONNode jsonJoint = jsonJoints[i];
			int jointType = jsonJoint["jointType"].AsInt;

			GameObject bodyA = loadedObjects[jsonJoint["bodyA"].AsInt];
			GameObject bodyB = loadedObjects[jsonJoint["bodyB"].AsInt];

			JSONNode localAnchorA = jsonJoint["localAnchorA"];
			Vector2 anchorA = new Vector2(localAnchorA[0].AsFloat / RATIO, -localAnchorA[1].AsFloat / RATIO);
			JSONNode localAnchorB = jsonJoint["localAnchorB"];
			Vector2 anchorB = new Vector2(localAnchorB[0].AsFloat / RATIO, -localAnchorB[1].AsFloat / RATIO);
			bool collideConnected = jsonJoint["collideConnected"].AsBool;
			string userData = jsonJoint["userData"].Value;

			if (jointType == (int) JointTypes.JOINT_DISTANCE || jointType == (int) JointTypes.JOINT_ROPE){
				DistanceJoint2D joint = bodyA.AddComponent<DistanceJoint2D>();
				joint.connectedBody = bodyB.GetComponent<Rigidbody2D>();
				joint.anchor = anchorA;
				joint.connectedAnchor = anchorB;

				// distance joint
				if (jsonJoint["length"] != null){
					joint.distance = jsonJoint["length"].AsFloat / RATIO;
					joint.maxDistanceOnly = true;
				}
				// rope joint
				else if (jsonJoint["maxLength"] != null){
					joint.distance = jsonJoint["maxLength"].AsFloat / RATIO;
				}

				joint.enableCollision = collideConnected;
				joint.name += '_';
				joint.name += userData.Length > 0 ? userData : "joint" + JointCount++;
			}
			else if (jointType == (int) JointTypes.JOINT_REVOLUTE){
				HingeJoint2D joint = bodyA.AddComponent<HingeJoint2D>();
				joint.connectedBody = bodyB.GetComponent<Rigidbody2D>();
				joint.anchor = anchorA;
				joint.connectedAnchor = anchorB;
				joint.enableCollision = collideConnected;
				joint.name += '_';
				joint.name += userData.Length > 0 ? userData : "joint" + JointCount++;

				// limits are not working properly 
				bool enableLimits = jsonJoint["enableLimit"].AsBool;
				float referenceAngle = -jsonJoint["referenceAngle"].AsFloat;
				float angleBetweenBodies = Mathf.Atan2(bodyB.transform.position.y - bodyA.transform.position.y,
				                                      bodyB.transform.position.x - bodyA.transform.position.x) * 180 / Mathf.PI;
				float upperAngle = -jsonJoint["lowerAngle"].AsFloat;
				float lowerAngle = -jsonJoint["upperAngle"].AsFloat;
				bool enableMotor = jsonJoint["enableMotor"].AsBool;
				float motorSpeed = -jsonJoint["motorSpeed"].AsFloat;
				float maxMotorTorque = jsonJoint["maxMotorTorque"].AsFloat;

				joint.useLimits = enableLimits;
				JointAngleLimits2D limits = new JointAngleLimits2D();
				limits.max = angleBetweenBodies + upperAngle;
				limits.min = angleBetweenBodies + lowerAngle;
				joint.limits = limits;
				joint.useMotor = enableMotor;
				JointMotor2D motor = new JointMotor2D();
				motor.maxMotorTorque = maxMotorTorque;
				motor.motorSpeed = motorSpeed;
				joint.motor = motor;
			}
			else if (jointType == (int) JointTypes.JOINT_WHEEL){
				WheelJoint2D joint = bodyA.AddComponent<WheelJoint2D>();
				joint.connectedBody = bodyB.GetComponent<Rigidbody2D>();
				joint.anchor = anchorA;
				joint.connectedAnchor = anchorB;
				joint.enableCollision = collideConnected;
				joint.name += '_';
				joint.name += userData.Length > 0 ? userData : "joint" + JointCount++;

				bool enableMotor = jsonJoint["enableMotor"].AsBool;
				float motorSpeed = -jsonJoint["motorSpeed"].AsFloat;
				float maxMotorTorque = jsonJoint["maxMotorTorque"].AsFloat;
				float dampingRatio = jsonJoint["dampingRatio"].AsFloat;
				float frequency = jsonJoint["frequencyHZ"].AsFloat;
				JSONNode localAxisA = jsonJoint["localAxisA"];
				float angle = Mathf.Atan2(-localAxisA[1].AsFloat, localAxisA[0].AsFloat) * 180 / Mathf.PI;

				joint.useMotor = enableMotor;
				JointMotor2D motor = new JointMotor2D();
				motor.maxMotorTorque = maxMotorTorque;
				motor.motorSpeed = motorSpeed;
				joint.motor = motor;

				JointSuspension2D suspension = new JointSuspension2D();
				suspension.dampingRatio = dampingRatio;
				suspension.frequency = frequency;
				suspension.angle = angle;
				joint.suspension = suspension;
			}
			else if (jointType == (int) JointTypes.JOINT_PRISMATIC){
				SliderJoint2D joint = bodyA.AddComponent<SliderJoint2D>();
				joint.connectedBody = bodyB.GetComponent<Rigidbody2D>();
				joint.anchor = anchorA;
				joint.connectedAnchor = anchorB;
				joint.enableCollision = collideConnected;
				joint.name += '_';
				joint.name += userData.Length > 0 ? userData : "joint" + JointCount++;
				
				bool enableLimits = jsonJoint["enableLimit"].AsBool;
				float referenceAngle = -jsonJoint["referenceAngle"].AsFloat;
				float upperTranslation = jsonJoint["upperTranslation"].AsFloat / RATIO;
				float lowerTranslation = jsonJoint["lowerTranslation"].AsFloat / RATIO;
				bool enableMotor = jsonJoint["enableMotor"].AsBool;
				float motorSpeed = -jsonJoint["motorSpeed"].AsFloat;
				float maxMotorTorque = jsonJoint["maxMotorTorque"].AsFloat;
				JSONNode localAxisA = jsonJoint["localAxisA"];
				float angle = Mathf.Atan2(-localAxisA[1].AsFloat, localAxisA[0].AsFloat) * 180 / Mathf.PI;

				joint.useLimits = enableLimits;
				JointTranslationLimits2D limits = new JointTranslationLimits2D();
				limits.max = upperTranslation;
				limits.min = lowerTranslation;
				joint.limits = limits;
				joint.useMotor = enableMotor;
				JointMotor2D motor = new JointMotor2D();
				motor.maxMotorTorque = maxMotorTorque;
				motor.motorSpeed = motorSpeed;
				joint.motor = motor;
				joint.angle = angle;
			}
		}
	}

	void loadJsonBodies(JSONNode jsonBodies){
		int BodyCount = 0;

		for (int i = 0, numberOfBodies = jsonBodies.Count; i < numberOfBodies; i++){
			JSONNode jsonBody = jsonBodies[i];
			
			int bodyType = jsonBody["type"].AsInt;
			JSONNode pos = jsonBody["position"];
			Vector3 position = new Vector3(pos[0].AsFloat / RATIO, -pos[1].AsFloat / RATIO, 0);
			float rotation = -jsonBody["rotation"].AsFloat;
			float linearDamping = jsonBody["linearDamping"].AsFloat;
			float angularDamping = jsonBody["angularDamping"].AsFloat;
			string userData = jsonBody["userData"].Value;
			bool isFixedRotation = jsonBody["isFixedRotation"].AsBool;
			bool isBullet = jsonBody["isBullet"].AsBool;
			
			
			GameObject body = new GameObject(userData.Length > 0 ? userData : "body" + BodyCount++);
			body.transform.position = position;
			body.transform.rotation = Quaternion.Euler(0, 0, rotation);
			body.AddComponent<DebugRenderer>();
			
			float density = 0;
			List<Fixture> fixtures = loadJsonFixtures(jsonBody["fixtures"]);
			for (int j = 0, numberOfFixtures = fixtures.Count; j < numberOfFixtures; j++){
				Fixture fixture = fixtures[j];
				density += fixture.density;
				if (fixture.shapeType == (int) ShapeTypes.SHAPE_BOX){
					BoxCollider2D boxCollider = body.AddComponent<BoxCollider2D>();
					boxCollider.isTrigger = fixture.isTrigger;
					boxCollider.offset = fixture.offset;
					boxCollider.size = fixture.size;
					boxCollider.sharedMaterial = fixture.physicsMaterial;
				}
				else if (fixture.shapeType == (int) ShapeTypes.SHAPE_CIRCLE){
					CircleCollider2D circleCollider = body.AddComponent<CircleCollider2D>();
					circleCollider.isTrigger = fixture.isTrigger;
					circleCollider.offset = fixture.offset;
					circleCollider.radius = fixture.size.x;
					circleCollider.sharedMaterial = fixture.physicsMaterial;
				}
				else if (fixture.shapeType == (int) ShapeTypes.SHAPE_POLYGON){
					PolygonCollider2D polyCollider = body.AddComponent<PolygonCollider2D>();
					polyCollider.isTrigger = fixture.isTrigger;
					polyCollider.offset = fixture.offset;
					polyCollider.SetPath(0, fixture.vertices);
					polyCollider.sharedMaterial = fixture.physicsMaterial;
				}
				else if (fixture.shapeType == (int) ShapeTypes.SHAPE_CHAIN){
					EdgeCollider2D edgeCollider = body.AddComponent<EdgeCollider2D>();
					edgeCollider.isTrigger = fixture.isTrigger;
					edgeCollider.offset = fixture.offset;
					edgeCollider.points = fixture.vertices;
					edgeCollider.sharedMaterial = fixture.physicsMaterial;
				}
			}

			body.AddComponent<Rigidbody2D>();
			Rigidbody2D rigidBody2D = body.GetComponent<Rigidbody2D>();
			rigidBody2D.isKinematic = bodyType == 1 || bodyType == 0;
			rigidBody2D.fixedAngle = isFixedRotation;
			rigidBody2D.mass = density;
			rigidBody2D.angularDrag = angularDamping;
			rigidBody2D.drag = linearDamping;

			if (isBullet){
				rigidBody2D.collisionDetectionMode = CollisionDetectionMode2D.Continuous;
			}
			else {
				rigidBody2D.collisionDetectionMode = CollisionDetectionMode2D.None;
			}

			loadedObjects.Add(body);
		}
	}

	List<Fixture> loadJsonFixtures(JSONNode jsonFixtures){
		List<Fixture> fixtures = new List<Fixture>();

		for (int i = 0, numberOfFixtures = jsonFixtures.Count; i < numberOfFixtures; i++){
			JSONNode jsonFixture = jsonFixtures[i];

			float density = jsonFixture["density"].AsFloat;
			float friction = jsonFixture["friction"].AsFloat;
			float restitution = jsonFixture["restitution"].AsFloat;
			bool isSensor = jsonFixture["isSensor"].AsBool;

			JSONNode jsonShapes = jsonFixture["shapes"];
			for (int j = 0, numberOfShapes = jsonShapes.Count; j < numberOfShapes; j++){
				JSONNode jsonShape = jsonShapes[j];

				JSONNode pos = jsonShape["position"];
				Vector2 position = new Vector2(pos[0].AsFloat / RATIO, -pos[1].AsFloat / RATIO);
				int shapeType = jsonShape["type"].AsInt;

				Fixture fixture = new Fixture();
				fixture.physicsMaterial = new PhysicsMaterial2D();
				fixture.physicsMaterial.friction = friction;
				fixture.physicsMaterial.bounciness = restitution;
				fixture.density = density;

				if (shapeType == (int) ShapeTypes.SHAPE_BOX){
					float width = jsonShape["width"].AsFloat / RATIO;
					float height = jsonShape["height"].AsFloat / RATIO;
					fixture.shapeType = (int) ShapeTypes.SHAPE_BOX;
					fixture.size = new Vector2(width, height);
					fixture.offset = position;
				}
				else if (shapeType == (int) ShapeTypes.SHAPE_CIRCLE){
					float radius = 2 * jsonShape["radius"].AsFloat / RATIO;
					fixture.shapeType = (int) ShapeTypes.SHAPE_CIRCLE;
					fixture.size = new Vector2(radius, radius);
					fixture.offset = position;
				}
				else if (shapeType == (int) ShapeTypes.SHAPE_POLYGON){
					JSONNode jsonVertices = jsonShape["vertices"];
					fixture.vertices = new Vector2[jsonVertices.Count];
					for (int k = 0, numberOfVertices = jsonVertices.Count; k < numberOfVertices; k++){
						fixture.vertices[k] = new Vector2(jsonVertices[k][0].AsFloat / RATIO, 
						                                  -jsonVertices[k][1].AsFloat / RATIO);
					}
					fixture.shapeType = (int) ShapeTypes.SHAPE_POLYGON;
					fixture.offset = position;
				}
				else if (shapeType == (int) ShapeTypes.SHAPE_CHAIN){
					JSONNode jsonVertices = jsonShape["vertices"];
					fixture.vertices = new Vector2[jsonVertices.Count];
					for (int k = 0, numberOfVertices = jsonVertices.Count; k < numberOfVertices; k++){
						fixture.vertices[k] = new Vector2(jsonVertices[k][0].AsFloat / RATIO, 
						                                  -jsonVertices[k][1].AsFloat / RATIO);
					}
					fixture.shapeType = (int) ShapeTypes.SHAPE_CHAIN;
					fixture.offset = position;
				}
				fixture.isTrigger = isSensor;
				fixtures.Add(fixture);
			}

		}

		return fixtures;
	}
}
